### PR TITLE
[RFC] Use pydantic

### PR DESCRIPTION
**Pull request by gadomski**
## Overview

This is a total rewrite of the library to use **pydantic** under the hood for validation and input/output schema generation. Key points:

- All tasks are now pydantic models, and their input and output data structures are also pydantic models. This allows jsonschema generation for input, output, and task config.
- Custom subclasses of `Task` are provided for common use-cases

| class definition | key method |
| -- | -- |
| `Task` | `process(self, input: List[Any]) -> List[Any]` |
| `StacOutTask` | `process_to_items(self, input: List[Any]) -> List[Item]` |
| `StacInStacOutTask` | `process_items(self, input: List[Item]) -> List[Item]` |
| `OneToManyTask` | `process_one_to_many(self, input: Any) -> List[Any]` |
| `OneToOneTask` | `process_one_to_one(self, input: Any) -> Any` |
| `ToItemTask` | `process_to_item(self, input: Any) -> Item` |
| `ItemTask` | `process_item(self, item: Item) -> Item` |
| `HrefTask` | `process_href(self, href: str) -> Item` |

- Models are also included for `Payload` and its substructures. This will hopefully solidify the data model for payloads (there had been confusion in the past around dict-or-list for process, etc)
- Handling has been moved up to `Payload`, so tasks just worry about themselves
- I'm using **stac-asset** instead of **fsspec** for read (and eventually, write)
- An example is provided of how to make a simple item using **rio-stac**, both as a standalone file and as a package. See the `examples/` directory for more information.
- CLI provides `stac-task list` for showing available tasks, and `stac-task jsonschema TASK MODEL` for dumping the jsonschema
- I've changed the package name to `stac_task` (from `stactask`) -- I figure this is a good chance to bring the naming in line with other repos AND clearly signal these very breaking changes to downstreams

Opening as a draft PR for visibility and feedback.

## Task list

- [x] Flesh out documentation
- [x] Unit test every generic class
- [x] Get the plugin story sorted out
- [x] Provide more examples of building custom tasks and how they would be called / invoked
- [ ] Asset download/upload helpers
  - As an aside, I'm not sure how much of this we actually want to put in **stac-task**, and how much should be farmed out to **stac-asset**
- [ ] Provide some common "helper" tasks, such as a **pystac-client** search. Maybe even include **rio-stac** via an extra dependencies set.
- [ ] Update Github branch protection for the new CI job names
- [ ] Build and deploy documentation with read the docs

### CODE REVIEW

File: src/stac_task/models.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+
+    path_template: str = r"${collection}"
+    """A string template for specifying the location of uploaded assets.
+    
+    See
+    https://pystac.readthedocs.io/en/stable/api/layout.html#pystac.layout.LayoutTemplate
+    for the available fields.
+    """
+
+    public_assets: Optional[Union[Literal["ALL"], List[str]]] = None
```
Comment (matthewhanson) [inline src/stac_task/models.py:32]: we removed the use of the public_assets, which is really specific to AWS.  
The upload options here are basically the options that should be passed through to stac-asset.

File: src/stac_task/models.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    """A list of asset keys that should be marked as public when uploaded"""
+
+    headers: Optional[Dict[str, str]] = None
+    """A set of key, value headers to send when uploading data to s3"""
+
+    collections: Optional[Dict[str, str]] = None
+    """A mapping of output collection name to a JSONPath pattern (for matching Items)"""
+
+    s3_urls: bool = False
```
Comment (matthewhanson) [inline src/stac_task/models.py:41]: Also specific to AWS, but not sure how best to specify something like this.  I think it depends on how stac-asset could eventually support uploads.

File: src/stac_task/payload.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        if len(matches) == 0:
+            raise ExecutionError("no tasks to execute")
+        elif len(matches) == 1:
+            key = matches.pop()
+            task_class = tasks[key]
+            task = task_class(**self.process.tasks[key])
+            task.payload_href = self.self_href
+            items = list()
+            # TODO async / thread pool
+            for item in self.features:
```
Comment (matthewhanson) [inline src/stac_task/payload.py:93]: A task doesn't work like this.  A process function takes in all the Items from the input and uses all of them to do something.  
Although I suppose this could be an additional feature that could add in another way to batch jobs...but that raises issues around state tracking and exception handling for batches I think are best avoided.

File: src/stac_task/task.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    """The output model."""
+
+    payload_href: Optional[str] = None
+    """The href of the payload that was used to execute this task."""
+
+    working_directory: str = Field(default_factory=TemporaryDirectory)
+    """The directory to save any downloaded files."""
+
+    @abstractmethod
+    def process(self, item: Input) -> List[Output]:
```
Comment (matthewhanson) [inline src/stac_task/task.py:35]: Handing in item (though should be ItemCollection) to the process function is way better than the current method which is to store the original items in the class.

I'm not sure if this needs to specify here that subclasses can define keywords, but a subclass can have keywords. The task parameters are then handed into the process function from those parameters.

Curious about other approaches but the advantage here is that it gives a task writer a simple experience - write the process function and any variables you need you add as keywords which will be automatically passed in via the payload.

File: examples/rio_stac_task.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
+import rio_stac.stac
+import stac_task
+from pystac import Item
+from stac_task import HrefTask
+
+
+class RioStacTask(HrefTask):
```
Comment (matthewhanson) [inline examples/rio_stac_task.py:7]: An example here on how I think this should be based on my other comments here (although here I'm using args and kwargs - required keywords are maybe better instead of args?)

```
class RioStacTask(StacOutputTask):
    def process(self,  href: str):
        return rio_stac.stac.create_stac_item(href)
```

File: src/stac_task/models.py (hunk 1 above)
Comment (gadomski) [reply src/stac_task/models.py:32]: Okay, I was doing my best to maintain backwards compatibility with the documented options. I'll remove this one though, or should we warn folks that we're ignoring it, if the field is present?

File: src/stac_task/models.py (hunk 2 above)
Comment (gadomski) [reply src/stac_task/models.py:41]: Yeah, the upload story in **stac-asset** is TBD and will likely mature along with this rewrite, so I'm inclined to keep this for now and then tweak it later once we figure out what **stac-asset** wants.

File: src/stac_task/payload.py (hunk 3 above)
Comment (gadomski) [reply src/stac_task/payload.py:93]: I was thinking about things from the developer's perspective, where each incoming item/href/assets is independent of the others. When I'm writing a `process` method, I only want to be thinking about one item at a time, right?

The handling of partial failures is best left to **stac-task**, which can collect and partition successful and unsuccessful calls to process. [We do something similar in **stac-asset**](https://github.com/stac-utils/stac-asset/blob/6c8f6c3a924fe51ba1e0c1d9d41d0b4195b6a577/src/stac_asset/_functions.py#L150-L165):

```python
        exceptions = list()
        for result in results:
            if isinstance(result, WrappedError):
                if self.config.error_strategy == ErrorStrategy.DELETE:
                    del result.download.owner.assets[result.download.key]
                else:
                    # Simple check to make sure we haven't added other
                    # strategies that we're not handling
                    assert self.config.error_strategy == ErrorStrategy.KEEP


                if self.config.warn:
                    warnings.warn(str(result.error), DownloadWarning)
                else:
                    exceptions.append(result.error)
        if exceptions:
            raise DownloadError(exceptions)
```

File: src/stac_task/task.py (hunk 4 above)
Comment (gadomski) [reply src/stac_task/task.py:35]: Keyword are passed on task instantiation, since they're constant for every call to `process`. A simple example from the tests: https://github.com/stac-utils/stac-task/blob/87bf2f02eb45c6d880a64f80bed76b018f31c4cb/tests/test_payload.py#L58-L63

You define the parameters for your task on the Task itself, which is a pydantic model: https://github.com/stac-utils/stac-task/blob/87bf2f02eb45c6d880a64f80bed76b018f31c4cb/tests/test_payload.py#L19-L28

File: examples/rio_stac_task.py (hunk 5 above)
Comment (gadomski) [reply examples/rio_stac_task.py:7]: See https://github.com/stac-utils/stac-task/pull/42#discussion_r1330322311, I think task parameters should be defined at task instantiation, not at process-time.

File: src/stac_task/models.py (hunk 1 above)
Comment (matthewhanson) [reply src/stac_task/models.py:32]: AFAIK no one is using it (none of our projects are).  In the cases where we are publishing public assets we've switched to setting the bucket policy, rather than using a per-Item ACL.
I think this PR is a big enough change that we don't need to worry about maintaining backward compatibility, but we should include a list of update notes of things that would need to change moving from old stac-task to the new version.

File: src/stac_task/payload.py (hunk 3 above)
Comment (matthewhanson) [reply src/stac_task/payload.py:93]: No, a process function is using all of the Items to create it's outputs.  Consider a data fusion example, where we are combining Landsat and Sentinel into a single derived product, or pan-sharpening.

Cirrus, and SWOOP is based on the idea of a single job having a single entry in the database. If you batch within a task then your state db does not represent a single job, it represents X jobs.  If Y of those fail and you want to rerun the others then those would end up going into the statedb leading to a mix of "batch" tasks and single item tasks.

In short, the orchestration platform is responsible for batching and orchestration, it should not be left to the tasks, but interested in @jkeifer's thoughts here.

File: src/stac_task/task.py (hunk 4 above)
Comment (matthewhanson) [reply src/stac_task/task.py:35]: Ok, so the payload parameters are used instantiate the task and are available as class members?  This seems weird to me as an old-schooler, defining keywords seems simpler, but whatever the Pydantic way is is fine.

File: examples/rio_stac_task.py (hunk 5 above)
Comment (matthewhanson) [reply examples/rio_stac_task.py:7]: I see now how the parameters are defined at instantiation and become class members. I think that's fine, just a change from the older paradigm I'm used to.

File: src/stac_task/payload.py (hunk 3 above)
Comment (gadomski) [reply src/stac_task/payload.py:93]: > No, a process function is using all of the Items to create it's outputs.

Still think this is an anti-pattern, because you can't define a schema for the `features` attribute of a payload, since it's a list. IMO the upstream "gathering" task should perform search-and-match, generating one item for each set of inputs that you want to fuse, e.g.:

```json
"features": [
  {
    "id": "item-0",
    "properties": {
       "start_datetime": "2023-09-19T00:00:00Z",
       "end_datetime": "2023-09-20T00:00:00Z"
    },
    "assets": {
       "landsat_band": { "you-get": "the-idea" },
       "sentinel_2_band": { }
    }
  },
  {
    "id": "item-1",
    "properties": {
       "start_datetime": "2023-09-18T00:00:00Z",
       "end_datetime": "2023-09-19T00:00:00Z"
    },
    "assets": {
      "landsat_band": {},
      "sentinel2_band": {}
    }
  }
]
```

That being said, the current architecture can easily switch to support your model. I'd just change the line we're commenting on to be:

```python
task.process_all(self.features)
```

and define a default `process_all` that does:

```python
def process_all(self, features: List[Input]) -> List[Output]:
    if len(features) != 1:
        raise NotImplemented("Cannot process of a list of more than one feature")
    else:
        return self.process(features[0])
```

Then, if tasks _did_ want to work on a list of (unschemad) inputs, they could just override `process_all`.

File: src/stac_task/task.py (hunk 4 above)
Comment (gadomski) [reply src/stac_task/task.py:35]: Nope, they're instance members, added to `__init__` by pydantic: https://github.com/stac-utils/stac-task/blob/87bf2f02eb45c6d880a64f80bed76b018f31c4cb/src/stac_task/payload.py#L89

I'm 👎🏼 on `**kwargs` because it requires a developer to read code (or possibly out-of-date documentation) to know values are supported and typed as, and allowing subclasses to add new arguments to an overriden method violates https://en.wikipedia.org/wiki/Liskov_substitution_principle (and makes **mypy** cranky).

File: src/stac_task/payload.py (hunk 3 above)
Comment (matthewhanson) [reply src/stac_task/payload.py:93]: It certainly makes data provenance a real problem, and puts a lot more logic in gathering together a payload into the feeder, it's pretty different then the original idea of how STAC workflows would work.  🤷

Comment (gadomski) [reply src/stac_task/payload.py:93]: > puts a lot more logic in gathering together a payload into the feeder, it's pretty different then the original idea of how STAC workflows would work. 

Logic's got to go somewhere, and IMO its better placed closer to the initial search, rather than separating the two.

> it's pretty different then the original idea of how STAC workflows would work. 

Fair enough, I am admittedly coming at this from a non-Cirrus perspective. Again, there's not much we'd have to change technically to support the system you describe, so we're not far off I don't think.

Comment (gadomski) [reply src/stac_task/payload.py]: Okay, I'm going to resolve this thread, since we now have `StacOutTask` and `StacInStacOutTask` to directly replicate the original task model.

**Comment by gadomski** [comment]
Closing as OBE. Maybe we'll re-visit someday.
